### PR TITLE
Live: throttle messages when FPS decreases

### PR DESCRIPTION
**Pull request by ryantxu**
hackish approach to avoid deadlock... parallel option for https://github.com/grafana/grafana/pull/32579

### CODE REVIEW

File: packages/grafana-runtime/src/measurement/fps.ts [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+  ok: true,
+};
+
+let lastUpdate = Date.now();
+function measure() {
+  let now = Date.now();
+  perf.fps = 1e3 / (now - lastUpdate);
+  lastUpdate = now;
+}
+setInterval(measure, 50); // 20hz
```
Comment (leeoniya) [inline packages/grafana-runtime/src/measurement/fps.ts:12]: i played with a few variants and this value should probably be some multiple of the framerate we want to reduce noise. it's not unusual for it to miss the deadline a bit on 1 cycle and catch up on the next. so here polling every 100ms and throttling when it exceeds 105ms works better for a 50ms target.

File: packages/grafana-runtime/src/measurement/perf.ts [hunk 2]
```diff
@@ -0,0 +1,27 @@
+let lastUpdate = Date.now();
+
+if (!(window as any).grafanaStreamingPerfBudget) {
+  (window as any).grafanaStreamingPerfBudget = 1.05;
```
Comment (ryantxu) [inline packages/grafana-runtime/src/measurement/perf.ts:4]: good idea?

This lets us experiment with different values easily in a production build..

image

File: packages/grafana-data/src/dataframe/StreamingDataFrame.ts [hunk 3]
```diff
@@ -42,7 +42,7 @@ function circPush(data: number[][], newData: number[][], maxLength = Infinity, d
     sliceIdx = nlen - maxLength;
   }
 
-  if (maxDelta !== Infinity) {
+  if (maxDelta !== Infinity && deltaIdx >= 0) {
```
Comment (ryantxu) [inline packages/grafana-data/src/dataframe/StreamingDataFrame.ts:45]: I keep getting burned by this crashing when -1 (no time field found yet)

Comment (leeoniya) [reply packages/grafana-data/src/dataframe/StreamingDataFrame.ts:45]: 👍

File: packages/grafana-runtime/src/measurement/perf.ts (hunk 2 above)
Comment (leeoniya) [reply packages/grafana-runtime/src/measurement/perf.ts:4]: ok for temp, not so sure about permanent. does it have to be on window or is there already some window._gfConfig exposed that we can go in one level. reading from globals in a perf sensitive loop makes me nervous.

File: packages/grafana-runtime/src/measurement/perf.ts [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+export const perf = {
+  budget: 1,
+  threshold: 1.05, // trial and error appears about right
+  ok: true,
+  last: lastUpdate,
+};
+
+// Expose this as a global object so it can be changed locally
+// NOTE: when we are confident this is the right budget, this should be removed
+(window as any).grafanaStreamingPerf = perf;
```
Comment (ryantxu) [inline packages/grafana-runtime/src/measurement/perf.ts:15]: @leeoniya better?

Comment (leeoniya) [reply packages/grafana-runtime/src/measurement/perf.ts:15]: lemme test if the loop is still 1ms / 10s...

**Review by leeoniya — APPROVED** [review]

File: packages/grafana-runtime/src/measurement/perf.ts (hunk 4 above)
Comment (leeoniya) [reply packages/grafana-runtime/src/measurement/perf.ts:15]: :+1:

**Pull request merged**